src/west/manifest.py

```diff
@@ -48,9 +54,229 @@
+    except (configparser.NoOptionError, configparser.NoSectionError) as e:
+        raise MalformedConfig('no "manifest.path" config option is set') from e
+
+# Manifest import handling
+
+def _default_importer(project: 'Project', file: str):
+    raise ManifestImportFailed(project, file)
+
+def _manifest_content_at(project: 'Project', path: str,
+                         rev: str = QUAL_MANIFEST_REV_BRANCH):
+    # Get a list of manifest data from project at path
+    #
+    # The data are loaded from Git at ref QUAL_MANIFEST_REV_BRANCH,
+    # *NOT* the file system.
+    #
+    # If path is a tree at that ref, the contents of the YAML files
+    # inside path are returned, as strings. If it's a file at that
+    # ref, it's a string with its contents.
+    #
+    # Though this module and the "west update" implementation share
+    # this code, it's an implementation detail, not API.
+
+    _logger.debug(f'{project.name}: looking up path {path} type at {rev}')
+
+    # Returns 'blob', 'tree', etc. for path at revision, if it exists.
+    out = project.git(['ls-tree', rev, path], capture_stdout=True,
+                      capture_stderr=True).stdout
+
+    if not out:
+        # It's a bit inaccurate to raise FileNotFoundError for
+        # something that isn't actually file, but this is internal
+        # API, and git is a content addressable file system, so close
+        # enough!
+        raise OSError(errno.ENOENT, os.strerror(errno.ENOENT), path)
+
+    ptype = out.decode('utf-8').split()[1]
+
+    if ptype == 'blob':
+        # Importing a file: just return its content.
+        return project.read_at(path, rev=rev).decode('utf-8')
+    elif ptype == 'tree':
+        # Importing a tree: return the content of the YAML files inside it.
+        ret = []
+        # Use a PurePosixPath because that's the form git seems to
+        # store internally, even on Windows. This breaks on Windows if
+        # you use PurePath.
+        pathobj = PurePosixPath(path)
+        for f in filter(_is_yml, project.listdir_at(path, rev=rev)):
+            ret.append(project.read_at(str(pathobj / f),
+                                       rev=rev).decode('utf-8'))
+        return ret
+    else:
+        raise MalformedManifest(f"can't decipher project {project.name} "
+                                f'path {path} revision {rev} '
+                                f'(git type: {ptype})')
+
+_import_map = collections.namedtuple('_import_map',
+                                     'file '
+                                     'name_whitelist path_whitelist '
+                                     'name_blacklist path_blacklist')
+
+# Type for an import map filter function, which takes a Project and
+# returns a bool. The various whitelists and blacklists are used to
+# create these filter functions. A None value is treated as a function
+# which always returns True.
+_ImapFilterFnType = Optional[Callable[['Project'], bool]]
```

Comment:
Would probably be nice to group custom types to the top or in a module.

Reply:
Is that a common convention? I was trying to keep the type definitions close to their first point of use. Happy to go either way.

Reply:
I understand the decision, fair enough. Afaik there's no rule, I've even seen some codebases to have a module with shared types, e.g. https://github.com/marshmallow-code/marshmallow/blob/dev/src/marshmallow/types.py

Reply:
OK, let's keep this as-is for now and see if @carlescufi or @tejlmand have a preference then. Thanks for the pointer!

Reply:
no strong opinions.
I can live with whatever decision @mbolivar-nordic prefers.

but I do feel the code is easier to read, when custom type definitions are group together, simply because the indentation is more `clean`/uniform in the method definition part this way.